zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutTrigger.java

```diff
@@ -75,12 +83,27 @@ private final class DeactivateTimeOutJobs implements Task {
 
     @Override
     public TaskResult execute(final TaskResultBuilder taskResultBuilder) {
-      final long now = currentTimeMillis();
-      state.forEachTimedOutEntry(
-          now,
-          (key, record) -> taskResultBuilder.appendCommandRecord(key, JobIntent.TIME_OUT, record));
+      if (startingTimestamp == -1) {
+        startingTimestamp = currentTimeMillis();
+      }
+
+      final Index nextIndex =
+          state.forEachTimedOutEntry(
+              startingTimestamp,
+              startAtIndex,
+              batchLimit,
+              (key, record) ->
+                  taskResultBuilder.appendCommandRecord(key, JobIntent.TIME_OUT, record));
+
       if (shouldReschedule) {
-        scheduleDeactivateTimedOutJobsTask();
+        if (nextIndex != null) {
+          startAtIndex = nextIndex;
+          scheduleDeactivateTimedOutJobsTask(Duration.ZERO);
+        } else {
+          startingTimestamp = -1;
+          startAtIndex = null;
+          scheduleDeactivateTimedOutJobsTask(pollingInterval);
+        }
```

Comment:
🔧 We could add some trace level logs every time it runs and debug level when it yields early. 

@megglos @mustafadagher What do you think?

Reply:
I'm fine with the trace logs, debug is too noisy IMO.

Reply:
:thought_balloon: just an observation, none of the other checkers in the engine log anything. was that some kind of a design decision?

Reply:
I don't think so, just something that we haven't bothered to do before. You can just copy how we do logging in other classes, nothing special for scheduled tasks necessary.